### PR TITLE
RavenDB-14254 Handling a conflict should take into account the unused…

### PR DESCRIPTION
**Pull request by karmeli87**
… databases

### CODE REVIEW

File: src/Raven.Server/ServerWide/Context/DocumentsOperationContext.cs [hunk 1]
```diff
@@ -12,7 +12,18 @@ public class DocumentsOperationContext : TransactionOperationContext<DocumentsTr
...
         internal Dictionary<string, long> LastReplicationEtagFrom;
-        internal bool SkipChangeVectorValidation;
+        private bool _skipChangeVectorValidation;
+
+        internal bool SkipChangeVectorValidation
+        {
+            get => _skipChangeVectorValidation;
+            set
+            {
+                if (_skipChangeVectorValidation == false)
```
Comment (arekpalinski) [inline src/Raven.Server/ServerWide/Context/DocumentsOperationContext.cs:22]: So we _never_ switch from `true` to `false` ?

File: src/Raven.Server/Documents/ConflictsStorage.cs [hunk 2]
```diff
@@ -872,6 +872,8 @@ public static ConflictStatus GetConflictStatusForDocument(DocumentsOperationCont
             else
                 return ConflictStatus.Update; //document with 'id' doesn't exist locally, so just do PUT
 
+            context.SkipChangeVectorValidation = context.DocumentDatabase.DocumentsStorage.TryRemoveUnusedIds(ref local);
```
Comment (ayende) [inline src/Raven.Server/Documents/ConflictsStorage.cs:875]: Why do we do this here?

File: src/Raven.Server/ServerWide/Context/DocumentsOperationContext.cs [hunk 3]
```diff
@@ -12,7 +12,18 @@ public class DocumentsOperationContext : TransactionOperationContext<DocumentsTr
 
         internal string LastDatabaseChangeVector;
         internal Dictionary<string, long> LastReplicationEtagFrom;
-        internal bool SkipChangeVectorValidation;
+        private bool _skipChangeVectorValidation;
+
+        internal bool SkipChangeVectorValidation
```
Comment (ayende) [inline src/Raven.Server/ServerWide/Context/DocumentsOperationContext.cs:17]: I'm not seeing usage of this?

File: src/Raven.Server/ServerWide/Context/DocumentsOperationContext.cs (hunk 1 above)
Comment (karmeli87) [reply src/Raven.Server/ServerWide/Context/DocumentsOperationContext.cs:22]: Not in the same transaction.

File: src/Raven.Server/ServerWide/Context/DocumentsOperationContext.cs (hunk 3 above)
Comment (karmeli87) [reply src/Raven.Server/ServerWide/Context/DocumentsOperationContext.cs:17]: This is for debug purpose to assert that we don't move the global change vector back in time.

File: src/Raven.Server/ServerWide/Context/DocumentsOperationContext.cs (hunk 1 above)
Comment (arekpalinski) [reply src/Raven.Server/ServerWide/Context/DocumentsOperationContext.cs:22]: I don't follow. Can you elaborate? What's purpose of it then?

Comment (karmeli87) [reply src/Raven.Server/ServerWide/Context/DocumentsOperationContext.cs:22]: This property is for debug assertion to validate that we do not moving the global change vector back in time (or conflicted). 

For example if it was `A:10 B:5` it shouldn't be set back to let's say `A:11` or `B:5`.
But there are some exceptions to this, for instance when we remove unused change vector entries. 
In this case we know that we will move the change vector back in time, so we need to skip the assertion.

We set the global change vector at the end of each transaction.
So once we set it to `true` we don't want to set it back to `false` until the transaction is over.

Comment (arekpalinski) [reply src/Raven.Server/ServerWide/Context/DocumentsOperationContext.cs:22]: Thanks. Makes sense

File: src/Raven.Server/Documents/ConflictsStorage.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/ConflictsStorage.cs:875]: @karmeli87 Can you answer this comment?

Comment (karmeli87) [reply src/Raven.Server/Documents/ConflictsStorage.cs:875]: We might put the the `local` one which after the removal of the entries is conflicted with the current global change vector, so we want to skip the assertion.

**Review by arekpalinski — APPROVED** [review]

**Comment by arekpalinski** [comment]
@ayende Could you check @karmeli87's answers and please approve if it's fine

**Review by ayende — APPROVED** [review]

**Pull request merged**